drgn_tools/meminfo.py

```diff
@@ -0,0 +1,755 @@
+
+
+def get_total_hugetlb_pages(prog: Program) -> int:
+    """Get sum of (4 kB) memory pages from all hstate hugepage pools."""
+    ret = 0
+    for h in for_each_hstate(prog):
+        pages_per_hugepage = 1 << (h.member_("order").value_())
+        ret += h.member_("nr_huge_pages").value_() * pages_per_hugepage
+    return ret
+
+
+def get_vm_commit_limit(prog: Program) -> int:
+    """
+    Get the total amount of memory (in pages) available to be allocated on the
+    system. Linux kernel may overcommit (i.e., allocating more than the amount
+    of physical memory). The result should consider |overcommit_kbytes| or
+    |overcommit_ratio|.
+    """
+    allowed = 0
+    total_pages = totalram_pages(prog).value_()
+    total_swap_pages = prog["total_swap_pages"].value_()
+    overcommit_kbytes = prog["sysctl_overcommit_kbytes"].value_()
+    overcommit_ratio = prog["sysctl_overcommit_ratio"].value_()
+
+    if overcommit_kbytes:
+        allowed = overcommit_kbytes >> (_PAGE_SHIFT - 10)
+    else:
+        hugetlb_pages = get_total_hugetlb_pages(prog)
+        allowed = (total_pages - hugetlb_pages) * overcommit_ratio // 100
+    allowed += total_swap_pages
+    return allowed
+
+
+def show_hugetlb_meminfo(prog: Program) -> None:
+    """Dump memory information for hugepages."""
+    hstate = prog["hstates"][prog["default_hstate_idx"]]
+    nr_huge_pages = hstate.member_("nr_huge_pages").value_()
+    free_huge_pages = hstate.member_("free_huge_pages").value_()
+    resv_huge_pages = hstate.member_("resv_huge_pages").value_()
+    surplus_huge_pages = hstate.member_("surplus_huge_pages").value_()
+    huge_page_order = hstate.member_("order").value_()
+    huge_page_size = 1 << (huge_page_order + _PAGE_SHIFT - 10)
+
+    print("HugePages_Total:   %5d" % (nr_huge_pages))
+    print("HugePages_Free:    %5d" % (free_huge_pages))
+    print("HugePages_Rsvd:    %5d" % (resv_huge_pages))
+    print("HugePages_Surp:    %5d" % (surplus_huge_pages))
+    print("Hugepagesize:   %8d kB" % (huge_page_size))
+
+    if _UEK_VERSION == 6 or _UEK_VERSION == 7:
+        total = 0
+        for h in for_each_hstate(prog):
+            count = h.member_("nr_huge_pages").value_()
+            total += (_PAGE_SIZE << huge_page_order) * count
+
+        print("Hugetlb:        %8d kB" % (total / 1024))
+
+
+def show_arch_meminfo(prog: Program) -> None:
+    """Dump numbers of pages mapped for supported page sizes."""
+    if _IS_AARCH_64:
+        return
+
+    direct_pages_count = prog["direct_pages_count"].read_()
+
+    direct_4k = direct_pages_count[prog.constant("PG_LEVEL_4K")].value_()
+    print("DirectMap4k:    %8d kB" % (direct_4k << 2))
+
+    if _IS_X86_64:
+        direct_2m = direct_pages_count[prog.constant("PG_LEVEL_2M")].value_()
+        print("DirectMap2M:    %8d kB" % (direct_2m << 11))
+    else:
+        direct_4m = direct_pages_count[prog.constant("PG_LEVEL_2M")].value_()
+        print("DirectMap4M:    %8d kB" % (direct_4m << 12))
```

Comment:
Is this for supporting x86 32 bit?

I wouldn't worry about that :)

Reply:
Ah, this is for non x86-64. I was not aware of that meminfo won't include DirectMap statistics for aarch64 when I implemented this function. Later, I added the above ``_IS_AARCH_64`` check and forgot to delete this part.
Deleted.